cpp/src/buffer/buffer.cpp

```diff
@@ -98,19 +153,19 @@ void buffer_copy(
         return;  // Nothing to copy.
     }
 
-    cuda_stream_join(std::array{dst.stream()}, std::array{src.stream()});
-    RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
-        dst.data() + dst_offset,
-        src.data() + src_offset,
-        size,
-        cudaMemcpyDefault,
-        dst.stream()
-    ));
-
-    // Override the event to track the async copy.
-    if (attach_cuda_event) {
-        src.override_event(CudaEvent::make_shared_record(src.stream()));
-        dst.override_event(CudaEvent::make_shared_record(dst.stream()));
-    }
+    // We have to sync both before *and* after the memcpy. Otherwise, `src.stream()`
+    // might deallocate `src` before the memcpy enqueued on `dst.stream()` has completed.
+    cuda_stream_join(std::views::single(dst.stream()), std::views::single(src.stream()));
+    dst.write_access(dst.stream(), [&](std::byte* dst_data) {
+        RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+            dst_data + dst_offset,
+            src.data() + src_offset,
+            size,
+            cudaMemcpyDefault,
+            dst.stream()
+        ));
+    });
+    cuda_stream_join(std::views::single(src.stream()), std::views::single(dst.stream()));
```

Comment:
Question: to my mind it is clearer what is going on if the `cuda_stream_join`s are _inside_ the function. Is that still correct?

i.e.

```
dst.write_access(dst.stream(), [&](std::byte *dst_data) {
   cuda_stream_join(dst, src);
   cudamemcpyasync(dst, src);
   cuda_stream_join(src, dst);
});
```